### PR TITLE
Implement support for mount options in PVs

### PR DESCRIPTION
**Pull request by gnufied**
**What this PR does / why we need it**:

This PR implements support for mount options in PersistentVolume via `volume.beta.kubernetes.io/mount-options` annotation.

**Which issue this PR fixes** 

Fixes https://github.com/kubernetes/features/issues/168

**Release note**:
```
Enable additional, custom mount options to be passed to PersistentVolume objects via volume.beta.kubernetes.io/mount-options annotation.
```

### CODE REVIEW

**Comment by rootfs** [comment]
Let's just add options to ReadWriteMany volume types (e.g. nfs and glusterfs) for now, and wait till provisioner is able to provision volumes and format into a desired fs type. Otherwise, PD/EBS/Cinder/etc  volumes could be formatted into either xfs or extX, the mount options won't be meaningful if fs type mismatches.

**Comment by gnufied** [comment]
@rootfs can you elaborate more? For attachable volume types - if file system on volume source and actual file system on device doesn't match - then mounting itself will fail regardless of supplied mount options.

The formatting is only attempted when device is not formatted to *any* file system type. And in that case - whatever mounting options specified in PV should work.

File: pkg/volume/util.go
Comment (saad-ali) [inline pkg/volume/util.go]: Please add unit tests for these methods.

File: pkg/volume/util.go
Comment (saad-ali) [inline pkg/volume/util.go]: Really what you need here is a set not a map. Maybe use https://github.com/kubernetes/kubernetes/tree/master/pkg/util/sets for clarity.

**Review by saad-ali — CHANGES_REQUESTED** [review]
As discussed offline, let's switch to annotation and target beta for 1.6

**Comment by redbaron** [comment]
what happens with this change if it is not ready before 1.6 freeze? will it be pushed to 1.7?

File: pkg/volume/util.go
Comment (gnufied) [inline pkg/volume/util.go]: fixed

File: pkg/volume/util/operationexecutor/operation_generator.go
Comment (gnufied) [inline pkg/volume/util/operationexecutor/operation_generator.go]: will do.

**Comment by humblec** [comment]
@rootfs @gnufied I can define the whiltelist/desired options for glusterfs.

**Comment by gnufied** [comment]
@humblec there is no need of that now. We got rid of whitelisting requirements. Admins can specify whatever they want in `mountOptions` for supported plugin types.

File: pkg/volume/util.go
Comment (gnufied) [inline pkg/volume/util.go]: done

File: pkg/api/validation/volume_plugins.go
Comment (rootfs) [inline pkg/api/validation/volume_plugins.go]: 2017

File: pkg/api/validation/volume_plugins.go
Comment (rootfs) [inline pkg/api/validation/volume_plugins.go]: consider consolidation with `ProbeVolumePlugins` in `cmd/kubelet/app/plugins.go`

**Comment by rootfs** [comment]
Consider when FSType can be set in provisioner from https://github.com/kubernetes/kubernetes/pull/40805 and mismatched mount option is set in the annotation, what's going to be solution?

**Comment by gnufied** [comment]
@rootfs Isn't it upto admin to make sure - he specifies mount option that works with his chosen fstype? Currently for example - if file system type on partition/disk and PV (specified via `VolumeSource`) doesn't matches then the mounting fails. The mount options themselves doesn't change this behaviour.

**Comment by saad-ali** [comment]
Looks good. Please squash your commits and update your original comment on this PR with a proper release-note, and I will LGTM/approve. Also looks like bazel is still failing.

**Comment by gnufied** [comment]
@saad-ali fixed up the PR description with proper text according to PR template. Also added release notes.

The last failure is in Jenkins verify:

```
I0224 17:55:45.529] +++ [0224 17:55:45] Stopping any currently running rsyncd container
I0224 17:55:45.656] +++ [0224 17:55:45] Starting rsyncd container
I0224 18:43:13.967] Terminate 16517 on timeout
E0224 18:43:14.069] Build timed out
E0224 18:43:14.069] Build failed
I0224 18:43:14.069] process 16517 exited with code -15 after 59.4m
```

Looks unrelated but I will keep an eye open.

**Comment by gnufied** [comment]
@k8s-bot kops aws e2e test this

**Comment by gnufied** [comment]
@k8s-bot gci gke e2e test this

**Comment by gnufied** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by gnufied** [comment]
@saad-ali the build looks green now, ptal

File: pkg/volume/flexvolume/flexvolume.go
Comment (eparis) [inline pkg/volume/flexvolume/flexvolume.go]: do not block merge on this comment, but isn't flex on where we should support mount options? Flex is the place we have absolutely no idea at all what people need, right?

Comment (childsb) [inline pkg/volume/flexvolume/flexvolume.go]: Since the mount is expected to be handled directly by the FLEX script, i dont think we want to take mount options in the FLEX volume plugin just yet.  We'd have to pass them through to the FLEX shell script mount command.  An easier solution if the storage admin/storage vendor wanted mount options they could edit their FLEX script directly to include them.

**Comment by gnufied** [comment]
Flaked on https://github.com/kubernetes/kubernetes/issues/40295 @k8s-bot bazel test this

**Comment by gnufied** [comment]
Flaked on https://github.com/kubernetes/kubernetes/issues/39471 @k8s-bot unit test this

**Comment by gnufied** [comment]
@k8s-bot cvm gce e2e test this

File: pkg/api/validation/volume_plugins.go
Comment (gnufied) [inline pkg/api/validation/volume_plugins.go]: I will do that post 1.6 merge I think. I didn't wanted to directly refer a function defined inside `cmd` package in `pkg/api` directly  - so I will put it in a common place or something like that.

File: pkg/volume/flexvolume/flexvolume.go
Comment (saad-ali) [inline pkg/volume/flexvolume/flexvolume.go]: Agree, would be a nice to have mount options piped to Flex. But like @childsb mentioned, it'll require updating the interface. Would be nice to get it into 1.6. Otherwise in 1.7 we could extend the Flex mount command parameters to accept mount options. +@chakri-nelluri

**Comment by saad-ali** [comment]
I tweaked your release-note a little. Otherwise LGTM

**Review by saad-ali — APPROVED** [review]

**Comment by smarterclayton** [comment]
/approve 

given general consensus

File: pkg/volume/plugins.go
Comment (smarterclayton) [inline pkg/volume/plugins.go]: This needs to be `mount-options`?

**Comment by smarterclayton** [comment]
Hold on LGTM, one API comment.

File: pkg/volume/plugins.go
Comment (gnufied) [inline pkg/volume/plugins.go]: We specced it to be `mountOptions` in design document.  I am open to changing the annotation. Is there a convention for this?

Comment (smarterclayton) [inline pkg/volume/plugins.go]: Yeah, as per:

> In fact, in-development API fields, including those used to represent fields of newer alpha/beta API versions in the older stable storage version, may be represented as annotations with the form something.alpha.kubernetes.io/name or something.beta.kubernetes.io/name (depending on our confidence in it). For example net.alpha.kubernetes.io/policy might represent an experimental network policy field. The "name" portion of the annotation should follow the below conventions for annotations. When an annotation gets promoted to a field, the name transformation should then be mechanical: foo-bar becomes fooBar.

Comment (smarterclayton) [inline pkg/volume/plugins.go]: > Key names should be all lowercase, with words separated by dashes, such as desired-replicas

Comment (gnufied) [inline pkg/volume/plugins.go]: fixed.

**Comment by smarterclayton** [comment]
Thanks, lgtm from API side

**Comment by gnufied** [comment]
Also updated release notes etc to reflect new `mount-options` string. I will open another PR to change the string in design proposal, but I don't think that should block this PR.

File: pkg/volume/plugins.go
Comment (saad-ali) [inline pkg/volume/plugins.go]: Good catch @smarterclayton

**Comment by saad-ali** [comment]
This PR and the ScaleIO PR (https://github.com/kubernetes/kubernetes/pull/38924) and the Portworx PR (https://github.com/kubernetes/kubernetes/pull/39535) will race: which ever ones get in first will force the others to to rebase. Just FYI

**Comment by gnufied** [comment]
I had to rebase this PR to accommodate master changes. cc @saad-ali @smarterclayton

**Comment by gnufied** [comment]
@k8s-bot test this

**Comment by gnufied** [comment]
@eparis @saad-ali had to rebase this PR because portworx got merged first. Needs another /lgtm

**Pull request merged**
